### PR TITLE
CA-345342: Made the query parameters of the http actions optional

### PR DESCRIPTION
**Pull request by kc284**
CA-345342: Made the query parameters of the http actions optional using as default value the type default.

Note that some of them are not optional, however, the current definition does not allow for distinguishing them from the mandatory ones, and as a consequence it is easier to mark them all as optional than provide method overloads manually.
Also, exposed manually in the SDK http actions removed from the API.
Use a mustache template to generate the HTTP actions.

### CODE REVIEW

**Comment by robhoes** [comment]
> using as default value the type default

Can you actually omit the parameters from the HTTP action, rather than filling in arbitrary defaults? I know that the C# bindings do this for RPCs as well, and it was hard to do it differently. However, in this case, it seems to be avoidable. I read something about adding the `[Optional]` attribute to the parameters rather than specifying a default, and then doing a null check in the function body.

File: csharp/gen_csharp_binding.ml [hunk 1]
```diff
@@ -174,89 +174,52 @@ and gen_relation out_chan (manyField, oneClass, oneField) =
...
-    | Varargs_query_arg -> "params string[] args /* alternate names & values */"
+    | Varargs_query_arg -> "params string[] args /* alternate names and values */"
   in
-
   let use_of_sdkarg =  function
-      String_query_arg s
+    | String_query_arg s
     | Int64_query_arg s
-    | Bool_query_arg s -> "\"" ^ s ^ "\", " ^ (escaped s)  (* "s", s *)
+    | Bool_query_arg s -> sprintf "\"%s\", %s" s (escaped s)
```
Comment (psafont) [inline csharp/gen_csharp_binding.ml:189]:
```replace
| Bool_query_arg s -> sprintf {|"%s", %s|} s (escaped s)
```

**Comment by kc284** [comment]
The default values provided are not arbitrary, they are the defaults for the specific type. If one replaces the explicit declaration with the `[Optional]` attribute, the compiler will still apply the type defaults, which means `false` and `0` for `bool` and `long` respectively, not `null`. The SDK does ommit the `null` and `false` args when constructing the URL query, but this leaves us indeed with the `0` for the `long` args. Perhaps what we should do here is change the `bool` and `long` args to nullable `bool?` and `long?`, so they can all accept `null` as the default.

**Comment by robhoes** [comment]
The defaults are well-defined for C#, but unrelated to the defaults applies by xapi. We may well define one of the Boolean args to have a default of `true`. If the SDK then omits `false` values, even if explicitly set, then you get the wrong behaviour. Having `null` as a default for all argument types, and then omit `null` args from the request, will avoid this.

**Review by robhoes — APPROVED** [review]
I think this looks good now.

**Review by AaronRobson — APPROVED** [review]

**Pull request merged**